src/mono/mono/metadata/class-internals.h

```diff
@@ -921,6 +921,8 @@ typedef struct {
 	MonoClass *generic_ienumerator_class;
 	MonoClass *alc_class;
 	MonoClass *appcontext_class;
+	MonoClass *weakreference_class;
+	MonoClass *generic_weakreference_class;
```

Comment:
nit. we don't need this in `mono_defaults`.
in sgen-mono.c` write:

```c
#include "metadata/class-internals.h"

...

static GENERATE_TRY_GET_CLASS_WITH_CACHE_DECL(weakreference);
static
GENERATE_TRY_GET_CLASS_WITH_CACHE_DECL(generic_weakreference);

GENERATE_TRY_GET_CLASS_WITH_CACHE (weakreference, "System", "WeakReference");
GENERATE_TRY_GET_CLASS_WITH_CACHE (generic_weakreference, "System", "WeakReference`1")
```

and then call `mono_class_try_get_weakreference_class` and `mono_class_try_get_generic_weakreference_class`

Reply:
Makes sense, I will update it.

Reply:
Actually, on second thought...

```
	/*
	Not using GENERATE_TRY_GET_CLASS_WITH_CACHE_DECL as this type is heavily checked by sgen when computing finalization.
	*/
	mono_defaults.critical_finalizer_object = mono_class_try_load_from_name (mono_defaults.corlib,
```

This is accessed just as often in the same code path and it was intentionally not done through `GENERATE_TRY_GET_CLASS_WITH_CACHE_DECL`. I can push the change to use the `GENERATE_*` methods but I want to double check first.

Reply:
`sgen_client_object_has_critical_finalizer` is called from multiple files, so generating an accessor method (assuming it's not done in a .h file) would make it not inlineable.
on the other hand WeakReference is (so far) only looked up from one .c file, and I'm suggesting putting the call to `GENEREATE` in the same .c file.  So we should get inlined code.

I don't feel very strongly about this, however. either way is an improvement over the status quo, I think

Reply:
I agree with @filipnavara that it wouldn't be safe to do class loading during a GC.